### ISSUE TITLE
[BUG] Photos lost on iPhone after app upgrade (Fails to create directory for Storage)

### ISSUE THREAD

**Issue by jsfan3** [codenameone/CodenameOne#2486]
I have an app that saves all its data in a TreeSet of PropertyBusinessObject. This data is saved to the Storage and it's restored from the Storage in the init(). Code:

```
public class StartPoint {

    private Form current;
    public static Resources theme;

    public static TreeMap<Long, ItemProperty> itemsMap = new TreeMap<>();

    public void init(Object context) {
        [...]
        Util.register("ItemProperty", ItemProperty.class);

        TreeMap restoredData = restoreData();
        if (restoredData != null) {
            itemsMap = restoredData;
            Log.p("Data restored");
        }
    }

    public static void saveData() {
        Storage.getInstance().writeObject("MyDiary", itemsMap);
        Log.p("Data saved");
    }

    public TreeMap restoreData() {
        HashMap map = (HashMap) Storage.getInstance().readObject("MyDiary");
        if (map != null) {
            return new TreeMap(map);
        } else {
            return null;
        }
    }
    
```

```
public class ItemProperty implements PropertyBusinessObject {

    public final Property<String, ItemProperty> photo = new Property<>("photo");
    public final LongProperty<ItemProperty> dateAndTime = new LongProperty<>("dateAndTime");
    public final Property<String, ItemProperty> quantity = new Property<>("quantity");
    public final Property<String, ItemProperty> place = new Property<>("place");
    public final Property<String, ItemProperty> feeling = new Property<>("feeling");
    public final Property<String, ItemProperty> fullness = new Property<>("fullness");
    public final Property<String, ItemProperty> afterEating = new Property<>("afterEating");
    public final Property<String, ItemProperty> other = new Property<>("other");

    private final PropertyIndex idx = new PropertyIndex(this, "ItemProperty",
            photo, dateAndTime, quantity, place, feeling, fullness,
            afterEating, other);

    @Override
    public PropertyIndex getPropertyIndex() {
        return idx;
    }

    public ItemProperty() {
        this.getPropertyIndex().registerExternalizable();
        Date currentTime = new Date();
        dateAndTime.set(currentTime.getTime());
    }

    public String getLocalizedDate() {
        Date date = new Date(dateAndTime.get());
        return L10NManager.getInstance().formatDateLongStyle(date);
    }
    
    public String getLocalizedDateTime() {
        Date date = new Date(dateAndTime.get());
        return L10NManager.getInstance().formatDateTime(date);
    }
    

    public String getTime() {
        Date date = new Date(dateAndTime.get());
        Calendar myCalendar = Calendar.getInstance();
        myCalendar.setTime(date);
        String hour = Integer.toString(myCalendar.get(Calendar.HOUR_OF_DAY));
        String minute = Integer.toString(myCalendar.get(Calendar.MINUTE));
        if (minute.length() == 1) {
            minute = "0" + minute;
        }
        return hour+":"+minute;
    }

}
```

I have an iPhone specific problem (that doesn't happen on Android). When I update the app **(installing a new build version)**, the photo of each ItemProperty is lost, while all the other data is correctly restored. Each photo was taken from the camera and then saved to the storage in this way:

```
// START OF IMAGE SECTION
        Style s = new Style();
        s.setFgColor(ColorUtil.GRAY);
        s.setBgColor(ColorUtil.WHITE);
        Label avatar = new Label(FontImage.createMaterial(FontImage.MATERIAL_ADD_A_PHOTO, s, imgSize / 3));
        avatar.getUnselectedStyle().setPadding(0, 0, 0, 0);
        avatar.getUnselectedStyle().setMargin(0, 0, 0, 0);

        if (data.photo.get() != null && !data.photo.get().isEmpty()) {
            try {
                Image img = Image.createImage(FileSystemStorage.getInstance().openInputStream(data.photo.get()));
                avatar.setIcon(img);
            } catch (IOException ex) {
                Log.p("ERROR: Cannot load the photo");
                Log.e(ex);
            }
        }

        ActionListener callback = e -> {
            Log.p("ActionListener callback invoked");
            if (e != null && e.getSource() != null) {
                String filePath = (String) e.getSource();

                // Resizes and Saves the image
                String pathAvatar = FileSystemStorage.getInstance().getAppHomePath() + System.currentTimeMillis() + ".jpg";
                try {
                    // About image orientation:
                    // https://stackoverflow.com/questions/51480978/codenameone-capture-photo-upside-down
                    OutputStream os2 = FileSystemStorage.getInstance().openOutputStream(pathAvatar);
                    Image img = Image.createImage(FileSystemStorage.getInstance().openInputStream(filePath));
                    if (img.getWidth() >= img.getHeight()) {
                        img = img.scaledWidth(Display.getInstance().convertToPixels(imgSize));
                    } else {
                        img = img.scaledHeight(Display.getInstance().convertToPixels(imgSize));
                    }
                    ImageIO.getImageIO().save(img, os2, ImageIO.FORMAT_JPEG, 0.9f);
                    os2.close();
                    // to open: Image img = Image.createImage(FileSystemStorage.getInstance().openInputStream(pathToImage));
                    data.photo.set(pathAvatar);
                    avatar.setIcon(img);
                    revalidate();
                } catch (IOException ex) {
                    Log.p("ERROR: Cannot save the photo");
                    Log.e(ex);
                }

            }
        };

        Style colorIcons = new Style();
        colorIcons.setFgColor(ColorUtil.LTGRAY);

        Button camera = new Button("Photocamera", FontImage.createMaterial(FontImage.MATERIAL_CAMERA, colorIcons, 5));
        camera.addActionListener((e) -> {
            Capture.capturePhoto(callback);
        });

        Label photo = new Label("Photo");
        photo.setAutoSizeMode(true);
        add(FlowLayout.encloseCenter(photo));
        Container imageAndButton = BoxLayout.encloseY(FlowLayout.encloseCenter(avatar), FlowLayout.encloseCenter(camera));
        imageAndButton.setLeadComponent(camera);
        add(imageAndButton);
        // END OF IMAGE SECTION
```

The iPhone log doesn't contain any interesting data to understand this issue, except this:

```
[EDT] 0:2:57,28 - 2018-07-27 11:42:20.177 StartPoint[354:19370] [LOG] 0:0:0,0 - Codename One revisions: c67824ee4d9359c0891df93ad3d03e6c349b5a3f
2018-07-27 11:42:20.181 StartPoint[354:19370] Failed to create directory /var/mobile/Containers/Data/Application/D0F5692E-2B39-4333-8AF1-32B8F83A291E/Documents/cn1storage/
```

I logged also the data saved before and after the app update. They are identical, they are in this form:
![json](https://user-images.githubusercontent.com/1997316/43317290-9387309c-919c-11e8-82de-1c1f92f557e8.png)

The strange fact is that "ERROR: Cannot load the photo" is not logged (see the code above) and the photos taken before the app upgrade are replaced by the placeholder. All the new photo are correctly saved and shown (even if I kill and restard the app)... until the next app upgrade, after that they are lost.

**Comment by codenameone** [comment]
Did you remove build hints from your project or change them? 
Specifically `ios.newStorageLocation`?

**Comment by jsfan3** [comment]
These are my build hints:
```

j2me.nativeThemeConst = 0
java.version = 8
ios.project_type = ios
win.ver = 8
ios.dsym = false
ios.prerendered_icon = false
ios.testFlight = false
rim.obfuscation = false
ios.statusbar_hidden = false
ios.application_exits = false
ios.includePush = true
ios.interface_orientation = UIInterfaceOrientationPortrait:UIInterfaceOrientationPortraitUpsideDown:UIInterfaceOrientationLandscapeLeft:UIInterfaceOrientationLandscapeRight
ios.NSCameraUsageDescription = Some functionality of the application requires your camera
ios.newStorageLocation = true
android.release = true
android.debug = false
```

**Comment by codenameone** [comment]
This doesn't relate to my original guess. Can you check if the files exist, what's their size etc?

You can use the `FileTree` class to look at the files you have in your file system and debug this on the device effectively.

**Comment by jsfan3** [comment]
It's unclear to me how to use `FileTree`, however I've just discovered that all saved images has file size of "0" kbytes...

I used this recursive method:

```
    /**
     * Logs the list of jpeg images, useful for debugging
     */
    private void getImagesList(String currentDir) throws IOException {
        if (currentDir == null) {
            for (String directory : FileSystemStorage.getInstance().getRoots()) {
                getImagesList(directory);
            }
        } else {
            for (String file : FileSystemStorage.getInstance().listFiles(currentDir)) {
                if (FileSystemStorage.getInstance().isDirectory(currentDir + file)) {
                    getImagesList(currentDir + file);
                } else if (file.endsWith("jpg")) {
                    Log.p("Found image: " + file
                            + " of size: "
                            + FileSystemStorage.getInstance().getLength(currentDir + file) / 1024
                            + " kbyte"
                    );
                }
            }
        }
    }
```

I got:

```
[LOG] 0:0:0,13 - -----------------------
[LOG] 0:0:0,13 - ---- LIST OF FILES ----
[LOG] 0:0:0,14 - Found image: temp_image.jpg of size: 4030 kbyte
[LOG] 0:0:0,15 - Found image: 1532750678563.jpg of size: 0 kbyte
[LOG] 0:0:0,15 - Found image: 1532725057980.jpg of size: 0 kbyte
[LOG] 0:0:0,16 - Found image: 1532640112119.jpg of size: 0 kbyte
[LOG] 0:0:0,16 - Found image: 1532711040333.jpg of size: 0 kbyte
[LOG] 0:0:0,16 - Found image: 1532897532647.jpg of size: 0 kbyte
[LOG] 0:0:0,17 - Found image: 1532684925691.jpg of size: 0 kbyte
[LOG] 0:0:0,17 - Found image: 1532788540075.jpg of size: 0 kbyte
[LOG] 0:0:0,17 - Found image: 1532538176679.jpg of size: 0 kbyte
[LOG] 0:0:0,18 - Found image: 1532532260765.jpg of size: 0 kbyte
[LOG] 0:0:0,18 - Found image: 1532559617099.jpg of size: 0 kbyte
[LOG] 0:0:0,19 - Found image: 1532615696676.jpg of size: 0 kbyte
[LOG] 0:0:0,19 - Found image: 1532632582036.jpg of size: 0 kbyte
[LOG] 0:0:0,19 - Found image: 1532535059954.jpg of size: 0 kbyte
[LOG] 0:0:0,20 - Found image: 1532682566742.jpg of size: 0 kbyte
[LOG] 0:0:0,20 - Found image: 1532696161408.jpg of size: 0 kbyte
[LOG] 0:0:0,22 - Found image: 1532684965822.jpg of size: 0 kbyte
[LOG] 0:0:0,22 - Found image: 1532858573492.jpg of size: 0 kbyte
[LOG] 0:0:0,23 - Found image: 1532717408336.jpg of size: 0 kbyte
[LOG] 0:0:0,23 - Found image: 1532847309772.jpg of size: 0 kbyte
[LOG] 0:0:0,23 - Found image: 1532586516490.jpg of size: 0 kbyte
[LOG] 0:0:0,24 - Found image: 1532926787555.jpg of size: 0 kbyte
[LOG] 0:0:0,24 - Found image: 1532623991754.jpg of size: 0 kbyte
[LOG] 0:0:0,24 - Found image: 1532862828302.jpg of size: 0 kbyte
[LOG] 0:0:0,25 - Found image: 1532596359355.jpg of size: 0 kbyte
[LOG] 0:0:0,25 - Found image: 1532532290362.jpg of size: 0 kbyte
[LOG] 0:0:0,26 - Found image: 1532543704860.jpg of size: 0 kbyte
[LOG] 0:0:0,26 - Found image: 1532538118685.jpg of size: 0 kbyte
[LOG] 0:0:0,26 - Found image: 1532761850383.jpg of size: 0 kbyte
[LOG] 0:0:0,27 - Found image: 1532682555037.jpg of size: 0 kbyte
[LOG] 0:0:0,27 - Found image: 1532603815602.jpg of size: 0 kbyte
[LOG] 0:0:0,27 - Found image: 1532777679317.jpg of size: 0 kbyte
[LOG] 0:0:0,28 - Found image: 1532796371117.jpg of size: 0 kbyte
[LOG] 0:0:0,28 - Found image: 1532552658354.jpg of size: 0 kbyte
[LOG] 0:0:0,29 - Found image: 1532678828633.jpg of size: 0 kbyte
[LOG] 0:0:0,29 - Found image: 1532933224567.jpg of size: 0 kbyte
[LOG] 0:0:0,29 - Found image: 1532891180157.jpg of size: 0 kbyte
[LOG] 0:0:0,30 - Found image: 1532690282596.jpg of size: 0 kbyte
[LOG] 0:0:0,30 - Found image: 1532750665070.jpg of size: 0 kbyte
[LOG] 0:0:0,30 - Found image: 1532897524188.jpg of size: 0 kbyte
[LOG] 0:0:0,31 - Found image: 1532812475912.jpg of size: 0 kbyte
[LOG] 0:0:0,31 - Found image: 1532870852415.jpg of size: 0 kbyte
[LOG] 0:0:0,31 - Found image: 1532803511068.jpg of size: 0 kbyte
```

**Comment by codenameone** [comment]
That's very weird. I made a fix a short while ago to a bug and I think my fix might have caused this regression. I've added a special guard against that. What I can't figure out is why this is happening as I can't see a path where this will occur. 

If you can build with include source for iOS (and have a Mac) you can test that this works by removing my original fix in xcode. If not it should be available by the weekend as usual and would hopefully solve this issue.

**Comment by jsfan3** [comment]
Ok. I haven't a Mac, so I'll wait that your commit will be in the build server on Friday.

**Comment by jsfan3** [comment]
Hi, I've just sent new builds of the apps which have this issue. Nothing is changed, the images are still 0kb, so the fix is failed.

I suppose that I reported enought code to implement a test case, however, if you prefer to have a ready fully implemented test case, I can publish on a Github repository the full code of a testing app.

**Comment by codenameone** [comment]
I'll try to reproduce a test case.

**Comment by codenameone** [comment]
I just created a test case and wasn't able to reproduce the issue. The file had a non-zero size on second run in an iPhone 6 device:

````java
Form test = new Form("Capture", BoxLayout.y());

Button t = new Button("Capture");
test.add(t);
t.addActionListener(e -> {
    String s = Capture.capturePhoto();
    if(s != null) {
        String pathAvatar = getAppHomePath() + System.currentTimeMillis() + ".jpg";
        try {
            OutputStream os2 = FileSystemStorage.getInstance().openOutputStream(pathAvatar);
            Image img = Image.createImage(FileSystemStorage.getInstance().openInputStream(s));
            if (img.getWidth() >= img.getHeight()) {
                img = img.scaledWidth(Display.getInstance().convertToPixels(50));
            } else {
                img = img.scaledHeight(Display.getInstance().convertToPixels(50));
            }
            ImageIO.getImageIO().save(img, os2, ImageIO.FORMAT_JPEG, 0.9f);
            os2.close();
            test.add(new Label(Image.createImage(pathAvatar)));
            test.revalidate();
        } catch (IOException ex) {
            Log.p("ERROR: Cannot save the photo");
            Log.e(ex);
        }
    }
});

try {
    String[] files = listFiles(getAppHomePath());
    for(String f : files) {
        long s = getFileLength(getAppHomePath() + f);
        test.add(new Label(f + " - " + s));
    }
} catch(IOException err) {
    Log.e(err);
}

test.show();
````

**Comment by jsfan3** [comment]
I spent two hours to try to understand why your test case works correctly without any issue (also on my iPhone X) and my app doesn't. The codes about the image capturing and storaging are very similar (quite identical). I also tried to remove the cn1libs and to insert the same build hints, but without success. It's very frustrating.

The only lesson that I learn in this story is that the next time that I write an app that uses the camera and the storage, I'll do an immediate check of the image file size, to discover any related bug just in time.

At the moment we can close this issue because I haven't a (simple) test case that reproduces the issue (except the whole app, but that it's not a test case).

However, thank you for the time you spent to try to solve this issue.
